xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,36 @@
 
 using namespace std;
 
-// The serialisation we use for doubles is inspired by a comp.lang.c post
-// by Jens Moeller:
-//
-// https://groups.google.com/group/comp.lang.c/browse_thread/thread/6558d4653f6dea8b/75a529ec03148c98
-//
-// The clever part is that the mantissa is encoded as a base-256 number which
-// means there's no rounding error provided both ends have FLT_RADIX as some
-// power of two.
-//
-// FLT_RADIX == 2 seems to be ubiquitous on modern UNIX platforms, while
-// some older platforms used FLT_RADIX == 16 (IBM machines for example).
-// FLT_RADIX == 10 seems to be very rare (the only instance Google finds
-// is for a cross-compiler to some TI calculators).
-
-#if FLT_RADIX == 2
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 7 + 7) / 8)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 8)
-# define MAX_MANTISSA (1 << (DBL_MAX_EXP & 7))
-#elif FLT_RADIX == 16
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 1 + 1) / 2)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 2)
-# define MAX_MANTISSA (1 << ((DBL_MAX_EXP & 1) * 4))
-#else
-# error FLT_RADIX is a value not currently handled (not 2 or 16)
-// # define MAX_MANTISSA_BYTES (sizeof(double) + 1)
-#endif
-
-static int base256ify_double(double &v) {
-    int exp;
-    v = frexp(v, &exp);
-    // v is now in the range [0.5, 1.0)
-    --exp;
-#if FLT_RADIX == 2
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
+string serialise_double(double v)
+{
+    size_t double_size = sizeof(double);
+#ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
+    memcpy(&temp, &v, double_size);
+    temp = do_bswap(temp);
+    return string(reinterpret_cast<const char *>(&temp), double_size);
 #endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
+    return string(reinterpret_cast<const char *>(&v), double_size);
```

Comment:
This really want to be in a `#else`.

Otherwise in the `WORDS_BIGENDIAN` case this line is dead code.  The compiler can spot that and not generate machine code for it, but it may also generate a warning about it.  Currently GCC doesn't, but e.g. clang does with `-Wunreachable-code-return`: "warning: 'return' will never be executed"

Reply:
Fixed in the latest commit.